### DISCUSSION TITLE
Defining new backends (akvirtualcamera)

### DISCUSSION THREAD

**Discussion by ratuspro** [letmaik/pyvirtualcam#87] [Q&A]
I've been playing with pyvirtualcam for the past couple of weeks and I'm currently facing the challenge of handling two virtual cameras that are not supported by the built in backends.

Unfortunately, I'm not able to figure out how to define in python other virtual cameras.
I've looked at the built in ones and they are all in cpp and there are no examples on how to create my custom cameras.

Btw, I'm using AkVCam to create new virtual devices.

**Comment by letmaik** [comment]
I think the word "defining" is not right here. It would be a new implementation for a new backend. And that typically requires writing native code. The existing backends written in C++ all follow the same structure, but obviously every backend needs different code to interact with a virtual camera. 

So, someone would have to write a backend that can talk to the akv virtual camera. I'm assuming you're referring to the macOS/Windows one from https://github.com/webcamoid/akvirtualcamera. How that code will look like and how easy it is will become clear at implementation time.

Any reason why you can't use the other available backends?

**Reply by ratuspro** [reply]
Hello @letmaik,

Thanks for the answer.

You are right, I would have to implement a new backend fro akv virtual cams.
From your answer, I'd figure it can only be done in c++ in a similar fashion to the built-in ones.
Would I need to build the entire package with an additional backend?

I think it is worth to try it out. In the end, it might make sense to do a PR for your repo.

I was not considering using the other available backends because for our use case, we might need more than 2 backends and we experienced some performance issues using Unity Capture backend.

**Reply by letmaik** [reply]
You don't have to implement the backend within the pyvirtualcam project itself. It can be an external package. Please see https://letmaik.github.io/pyvirtualcam/#pyvirtualcam.register_backend which shows the API interface to register a new backend. The idea is that the new backend defines a Python class that then exposes a few functions to interact with the virtual camera. This is how the built-in backends are registered: https://github.com/letmaik/pyvirtualcam/blob/56b6c73e6b796ccdb3bf36907380d54a4b5cbed3/pyvirtualcam/camera.py#L107-L116
The built-in backends use the C++ pybind11 library to implement/expose the `Backend` Python class that calls into other C++ code talking to the device. There is no fundamental requirement to use C++ though. If you can talk to the camera device efficiently from Python, then by all means implement it all in Python, but it's often not the best choice.

**Reply by letmaik** [reply]
Follow-up... I think extending pyvirtualcam (and PRing) may be easier than developing and maintaining a separate package. So I would probably start with a fork as it reduces some amount of work to get started. Having said that, I think I would only accept new backends written in C++ (to remain as performant as possible) following a similar structure to the existing ones (to avoid maintenance burden).

**Reply by ratuspro** [reply]
For the moment, I think I'll try to implement it using python as I already have some code for that.
Still, I'm not sure if it will work. I'll keep you updated. :)

I agree with extending pyvirtualcam with a c++ backend.
I expect to have one written so I might as well do a PR with it.

Btw, any suggestion on how to build the package locally besides the ones here https://github.com/letmaik/pyvirtualcam#windows ?

**Reply by letmaik** [reply]
Re building, it's mostly `python -u setup.py bdist_wheel` in a Python prompt, I typically use Anaconda as that comes with the native build toolchain (for C++ extensions). That creates an installable `.whl` package.

**Reply by ratuspro** [reply]
Sounds great! 
I'll drop some updates as soon as possible.

Thanks for all the help.

**Reply by ratuspro** [reply]
You may find my progress on the following [fork](https://github.com/ratuspro/pyvirtualcam) to add a AkVCam Backend.

As suggested, I'm taking inspiration from the other backends. 
Currently, the backend is able to find the devices in the registry as well as their preferences (height, width, format, etc...)

However, I'm experiencing some problems while trying to bind to the devices IO.
I'm following the examples provided on akvirtualcamera repo ([here](https://github.com/webcamoid/akvirtualcamera/blob/master/share/examples/output_windows.c)). They launch a new process to handle the new frames.

I'm having problems defining `struct StreamProcess` because, during the setup, some of its fields types are not found (e.g. 'SECURITY_ATTRIBUTES'). I'm pretty sure it has something to do with the libraries linked in extension modules. However, I'm not familiar with setuptools nor pybind11.

If you could provide some assistance I would greatly appreciate.

**Reply by letmaik** [reply]
According to the [docs](https://docs.microsoft.com/en-us/previous-versions/windows/desktop/legacy/aa379560(v=vs.85)) it should be enough to include `windows.h`. Not sure why you're getting errors there.

**Reply by ratuspro** [reply]
I also though the same. Well, after changing to miniconda it works.

As it stands, the AkVCam backend is capable of:
- identifying what virtual cameras are available on a windows machine
- start streaming using the AkVCam supported formats (still missing other formats available for other backends)

I've still not worked on test, docs, or examples.
I added you to my fork as a collaborator. 
If you prefer I can do a PR with this contribution and we can iterate on this new backend there.

**Reply by letmaik** [reply]
Great! Yes, I think a PR makes sense, then it makes it easier to review and add comments.
A few notes already here:
- The registry has the install path of akvcam, which should be used to construct the full exe path.
- Instead of C buffers, let's use `std::vector<uint8_t>`, then memory gets deallocated automatically
- Can a frame be written to stdin in full or does it have to be line by line? The former would be simpler
- The external process may exit at any point which needs to be handled
- It's fine to not support all image formats

**Reply by ratuspro** [reply]
Thanks for the feedback.

This can be found here https://github.com/letmaik/pyvirtualcam/pull/88
